### PR TITLE
Run OSD integ test ci-groups in parallel

### LINKED ISSUE #4944: Ensure the OpenSearch Dashboards `ci-group` tests run in parallel
### Is your feature request related to a problem? Please describe

Coming from https://github.com/opensearch-project/opensearch-dashboards-functional-test/pull/1411 and https://github.com/opensearch-project/opensearch-build/pull/4796 after adding the ci-groups, although the integration tests currently run in parallel for all components included in the input manifest, the CI groups are not running concurrently since they are all part of the single component, OpenSearch Dashboards.

### Describe the solution you'd like

Run all the ci-groups in parallel for component, OpenSearch Dashboards. Here is the sample build that took 6 plus hours https://build.ci.opensearch.org/job/integ-test-opensearch-dashboards/6248/ because the ci-groups are not running in parallel.

### Describe alternatives you've considered

_No response_

### Additional context

_No response_

### PR DESCRIPTION
**Pull request by rishabh6788**
### Description
Run OSD integ test ci-groups in parallel. 

This change proposes a dedicated job to run OSD integration test ci-groups in parallel. 
The existing job can add a check to remove OSD component from list and run as usual. 

### Issues Resolved
#4944 

By submitting this pull request, I confirm that my contribution is made under the terms of the Apache 2.0 license.
For more information on following Developer Certificate of Origin and signing off your commits, please check [here](https://github.com/opensearch-project/OpenSearch/blob/main/CONTRIBUTING.md#developer-certificate-of-origin).

### CODE REVIEW

File: src/test_workflow/integ_test/integ_test_runner.py [hunk 1]
```diff
@@ -43,14 +43,13 @@ def run(self) -> TestSuiteResults:
...
-                                component.name = f"{orig_component_name}-ci-group-{i}"
-                                test_suite = self.__create_test_suite__(component, test_config, work_dir.path)
-                                test_results = test_suite.execute_tests()
-                                [self.test_recorder.test_results_logs.generate_component_yml(result_data) for result_data in test_suite.result_data]
-                                all_results.append(component.name, test_results)
+                            component.name = f"{orig_component_name}-ci-group-{self.args.ci_group}"
+                            test_suite = self.__create_test_suite__(component, test_config, work_dir.path)
+                            test_results = test_suite.execute_tests()
+                            [self.test_recorder.test_results_logs.generate_component_yml(result_data) for result_data in test_suite.result_data]
+                            all_results.append(component.name, test_results)
```
Comment (gaiksaya) [inline src/test_workflow/integ_test/integ_test_runner.py:52]: Looks like if and else block has a duplicate code. Does the below recommendation makes sense? Only change component-name if ci-group is present. Everything goes on as before. 

```python
if self.args.ci_group:
    orig_component_name = component.name
    component.name = f"{orig_component_name}-ci-group-{self.args.ci_group}"

test_suite = self.__create_test_suite__(component, test_config, work_dir.path)
test_results = test_suite.execute_tests()
[self.test_recorder.test_results_logs.generate_component_yml(result_data) for result_data in test_suite.result_data]
all_results.append(component.name, test_results)

File: tests/tests_test_workflow/test_integ_workflow/integ_test/test_integ_test_runner_opensearch_dashboards.py [hunk 2]
```diff
@@ -149,8 +151,6 @@ def test_with_integ_test_ci_groups(self, mock_temp: Mock, mock_test_recorder: Mo
         results = runner.run()
 
         self.assertEqual(results["sql-ci-group-1"], mock_test_results)
-        self.assertEqual(results["sql-ci-group-2"], mock_test_results)
```
Comment (gaiksaya) [inline tests/tests_test_workflow/test_integ_workflow/integ_test/test_integ_test_runner_opensearch_dashboards.py:152]: Can we check his assertions for given ci-group number?

File: jenkins/opensearch-dashboards/dashboards-integ-test.jenkinsfile [hunk 3]
```diff
@@ -0,0 +1,376 @@
+/*
```
Comment (gaiksaya) [inline jenkins/opensearch-dashboards/dashboards-integ-test.jenkinsfile:1]: Can you please make changes in current integ-test workflow so that diff is easily visible?

File: tests/tests_test_workflow/test_integ_workflow/integ_test/test_integ_test_runner_opensearch_dashboards.py (hunk 2 above)
Comment (rishabh6788) [reply tests/tests_test_workflow/test_integ_workflow/integ_test/test_integ_test_runner_opensearch_dashboards.py:152]: This test is checking the assertion when the `ci-group` parameter is passed with value 1. 
I just deleted the previous implementation of running all ci-groups in the manifest file.

File: jenkins/opensearch-dashboards/dashboards-integ-test.jenkinsfile (hunk 3 above)
Comment (rishabh6788) [reply jenkins/opensearch-dashboards/dashboards-integ-test.jenkinsfile:1]: I am proposing a dedicated job for OSD integ test because adding this ci-group changes in the existing job will be hacky and not clean. Running ci-groups and different components are technically two different use-cases. 
This can be done in existing file but will make it clunky and hacky.

Comment (gaiksaya) [reply jenkins/opensearch-dashboards/dashboards-integ-test.jenkinsfile:1]: How about putting it in a library and calling the library instead. Seeing a lot of duplicate code and also worried about the maintenance like changing docker image names in both files, agent names, etc.

Comment (rishabh6788) [reply jenkins/opensearch-dashboards/dashboards-integ-test.jenkinsfile]: Refactored to use the same job file.

File: jenkins/opensearch-dashboards/integ-test.jenkinsfile [hunk 4]
```diff
@@ -230,73 +230,94 @@ pipeline {
...
-                        def wait_seconds = local_component_index * 20
 
-                        echo "Adding Component: ${local_component}"
-                        componentTests["Run Integtest ${local_component}"] = {
-                            // Using scripted pipelines to trigger dynamic parallel stages
-                            timeout(time: 6, unit: 'HOURS') {
+                        // Changes to run OpenSearch-Dashboards ci-groups in parallel.
+                        // There are total 9 ci-groups
+                        if (component == "OpenSearch-Dashboards") {
+                          for (int i = 1; i <= 9; i++) {
```
Comment (peterzhuamazon) [inline jenkins/opensearch-dashboards/integ-test.jenkinsfile:237]: Can we dynamically assign this value as workflow params?

Comment (rishabh6788) [reply jenkins/opensearch-dashboards/integ-test.jenkinsfile:237]: Yes, can take it up in future improvements.

Comment (gaiksaya) [reply jenkins/opensearch-dashboards/integ-test.jenkinsfile:237]: Can you please create an issue for this? Also wondering if we can dynamically assign this value by fetching from test-manifest instead of hard coding to 9 here.
TL;DR:
```
ci-number = ${ci_number} <------------ parameter to this jenkins file 
if ci_number == null:
  ci-number = testManifest.OpenSearch-Dashboards.ci_group  <----------- Parse test manifest to fetch this value dynamically
```

File: tests/jenkins/TestOpenSearchDashboardsIntegTest.groovy [hunk 5]
```diff
@@ -99,6 +100,15 @@ class TestOpenSearchDashboardsIntegTest extends BuildPipelineTest {
...
         helper.registerAllowedMethod('unstash', [String.class], null)
+        helper.registerAllowedMethod("parallel", [Map]) { stages -> 
+    println "Mock parallel stages:"
+    stages.each { stageName, stageContent ->
+        println "\nStage: ${stageName}"
+        println "Content: ${stageContent.toString()}"
+        // Execute the stage content to simulate parallel execution
+        stageContent.call()
+    }
+}
```
Comment (gaiksaya) [inline tests/jenkins/TestOpenSearchDashboardsIntegTest.groovy:111]: Good find!

**Review by gaiksaya — APPROVED** [review]
